client/network/src/request_responses.rs

```diff
@@ -0,0 +1,866 @@
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Collection of request-response protocols.
+//!
+//! The [`RequestResponses`] struct defined in this module provides support for zero or more
+//! so-called "request-response" protocols.
+//!
+//! A request-response protocol works in the following way:
+//!
+//! - For every emitted request, a new substream is open and the protocol is negotiated. If the
+//! remote supports the protocol, the size of the request is sent as a LEB128 number, followed
+//! with the request itself. The remote then sends the size of the response as a LEB128 number,
+//! followed with the response.
+//!
+//! - Requests have a certain time limit before they time out. This time includes the time it
+//! takes to send/receive the request and response.
+//!
+//! - If provided, a ["requests processing"](RequestResponseConfig::requests_processing) channel
+//! is used to handle incoming requests.
+//!
+
+use futures::{channel::{mpsc, oneshot}, prelude::*};
+use libp2p::{
+	core::{
+		connection::{ConnectionId, ListenerId},
+		ConnectedPoint, Multiaddr, PeerId,
+	},
+	request_response::{
+		RequestResponse, RequestResponseCodec, RequestResponseConfig, RequestResponseEvent,
+		RequestResponseMessage, ResponseChannel, ProtocolSupport
+	},
+	swarm::{
+		protocols_handler::multi::MultiHandler, NetworkBehaviour, NetworkBehaviourAction,
+		PollParameters, ProtocolsHandler,
+	},
+};
+use std::{
+	borrow::Cow, collections::{hash_map::Entry, HashMap}, io, iter, pin::Pin,
+	task::{Context, Poll}, time::Duration,
+};
+
+pub use libp2p::request_response::{InboundFailure, OutboundFailure, RequestId};
+
+/// Configuration for a single request-response protocol.
+#[derive(Debug, Clone)]
+pub struct ProtocolConfig {
+    /// Name of the protocol on the wire. Should be something like `/foo/bar`.
+    pub name: Cow<'static, str>,
+
+    /// Maximum allowed size, in bytes, of a request.
+    ///
+    /// Any request larger than this value will be declined as a way to avoid allocating too
+    /// much memory for it.
+    pub max_request_size: usize,
+
+    /// Maximum allowed size, in bytes, of a response.
+    ///
+    /// Any response larger than this value will be declined as a way to avoid allocating too
+    /// much memory for it.
+    pub max_response_size: usize,
```

Comment:
Maybe instead of using `usize` we should be machine independent and use `u32` or `u64` for `max_request_size` and `max_response_size`?